### ISSUE TITLE
ExitStack fails with StreamReader / StreamWriter

### ISSUE THREAD

**Issue by pganssle** [ncoghlan/contextlib2#5]
In Python 2.7, [`StreamReader`](https://docs.python.org/2/library/codecs.html#codecs.StreamReader) and [`StreamWriter`](https://docs.python.org/2/library/codecs.html#codecs.StreamWriter) are context managers that wrap around a stream and handle encoding issue. However, the `type` of one of these objects is `instance`, not `file`. As a result, `ExitStack.enter_context` with `AttributeError` when one of these is added to the stack:

MWE:
```python
# -*- coding: utf-8 -*-
from contextlib2 import ExitStack
from codecs import getreader

with open('test_file', 'wb') as f:
    f.write(u'いちりきつくっ'.encode('UTF-8'))

reader = getreader('utf-8')
with reader(open('test_file', 'r')) as f:
    print(f.read())  # いちりきつくっ

try:
    with ExitStack() as stack:
        f = stack.enter_context(reader(open('test_file', 'r'))) # Error
        print(f.read())
finally:
    import os
    os.unlink('test_file')    # Delete the file
```

Here is the result:

```
AttributeError: type object 'instance' has no attribute '__exit__'
```

Not sure what happens on Python 3, since Python 3's native unicode support obviates the need for the `StreamReader` in the first place as far as I can tell.

I don't know if this is a good solution, but if you use `cm.__class__` instead of `type(cm)`, it seems to work OK:

```python
    def enter_context(self, cm):
        """Enters the supplied context manager

        If successful, also pushes its __exit__ method as a callback and
        returns the result of the __enter__ method.
        """
        # We look up the special methods on the type to match the with statement
        for _cm_type in type(cm), cm.__class__:
            try:
                _exit = _cm_type.__exit__
                result = _cm_type.__enter__(cm)
            except AttributeError as e:
                error = e
                continue

            error = None
            break

        if error is not None:
            raise error        

        self._push_cm_exit(cm, _exit)
        return result
```

**Comment by ncoghlan** [comment]
The bug is that ExitStack expects new-style classes, which is true by default in Python 3, but needs to be requested explicitly in Python 2.7: https://github.com/ncoghlan/contextlib2/commit/e48fbea5a592229803120725e016be1006b85380

I just need to push out a 0.5.2 release with Devin's fix in it.

**Comment by ncoghlan** [comment]
Ah, sorry, I misread the issue: I thought you were talking about `redirect_stderr` (etc), which needed to be fixed in the backport itself (since they're new-style classes in Python 3, but were defaulting to old-style in Python 2).

For this, I think a better way to tackle the problem might be to find the code that does "type(cm)" (the case you found isn't the only place), and replace it with a `_get_type` helper function defined as follows:

    try:
        from types import ClassType
    except ImportError:
        _get_type = type # Python 3 doesn't have old-style classes
    else:
        def _get_type(obj):
            obj_type = type(obj)
            if obj_type is ClassType:
                return obj.__class__ # Old-style class
            return obj_type # New-style class

**Comment by ncoghlan** [comment]
The code above wasn't quite right, but I believe the commit I just pushed should fix the compatibility problem with old-style classes - are you able to test against master and see if that addresses the problem you found in https://github.com/larsyencken/cjktools/pull/5 ?

**Comment by pganssle** [comment]
@ncoghlan Yes, that solves the issue. Thanks!

**Comment by ncoghlan** [comment]
Thanks for the confirmation! Now published to PyPI as 0.5.3.
